### PR TITLE
Fetch account data with zstd to shrink payload size

### PR DESCRIPTION
**Pull request by jarry-xiao**

### CODE REVIEW

File: typescript/phoenix-sdk/src/client.ts [hunk 1]
```diff
@@ -139,26 +144,35 @@ export class Client {
       accountList.push(...marketAddresses);
     }
     accountList.push(SYSVAR_CLOCK_PUBKEY);
-    const accounts = await connection.getMultipleAccountsInfo(
-      accountList,
-      "confirmed"
-    );
-    const clockBuffer = accounts.pop()?.data;
+    let accounts: Buffer[] = [];
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:147]: `const accounts = useZstd ? await getAccountsZstd(...) : await getAccountsNormal(...)`

File: typescript/phoenix-sdk/src/client.ts [hunk 2]
```diff
@@ -202,15 +216,25 @@ export class Client {
...
+    useZstd = true
   ): Promise<Client> {
     const cluster = await getClusterFromConnection(connection);
-    const accounts = await connection.getMultipleAccountsInfo(
-      [...marketAddresses, SYSVAR_CLOCK_PUBKEY],
-      "confirmed"
-    );
+    let accounts: Buffer[];
+
+    if (useZstd) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:225]: same here

File: typescript/phoenix-sdk/src/client.ts [hunk 3]
```diff
@@ -323,9 +348,13 @@ export class Client {
    * Add a market to the client. Useful for localnet as markets will not be loaded in by default.
    * @param marketAddress The `PublicKey` of the market account
    * @param forceReload If this is set to true, it will reload the market even if it already exists
-   *
+   * @param useZstd If this is set to true, it will use zstd compression to get the market data
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:351]: add indication for when you would want to use this

File: typescript/phoenix-sdk/src/client.ts [hunk 4]
```diff
@@ -344,48 +373,57 @@ export class Client {
...
+      [marketBuffer, clockBuffer] =
+        await getConfirmedMarketsAndClockAccountsZstd(this.connection, [
+          marketKey,
+        ]);
+    } else {
+      [marketBuffer, clockBuffer] = await getConfirmedMarketsAndClockAccounts(
+        this.connection,
+        [marketKey]
+      );
     }
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:389]: `const [marketBuffer, clockBuffer] = useZstd ? ... : ...`

File: typescript/phoenix-sdk/src/utils/connection.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  const clockBuffer = Buffer.from(decoder.decode(compressedClockData));
+
+  const markets: Buffer[] = [];
+  for (const payload of values) {
+    if (payload?.data[0] === undefined) {
+      throw new Error("Unable to get market account data");
+    }
+    const compressedMarketData = Buffer.from(payload.data[0], "base64");
+    const marketBuffer = Buffer.from(
+      decoder.decode(compressedMarketData, 1 << 23)
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/connection.ts:72]: comment why 1 << 23

probably more readable as a const MAX_ACCOUNT_SIZE_BYTES

Comment (eugene-chen) [reply typescript/phoenix-sdk/src/utils/connection.ts:72]: (or MAX_MARKET_SIZE_BYTES, slightly prefer the former)

File: typescript/phoenix-sdk/src/utils/connection.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  const response = await fetch(connection.rpcEndpoint, {
+    method: "POST",
+    body: JSON.stringify(body),
+    headers: {
+      "Content-Type": "application/json",
+    },
+  });
+
+  const value = (await response.json()).result.value;
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/connection.ts:109]: you really need to await response.json() after awaiting the fetch?

File: typescript/phoenix-sdk/src/client.ts [hunk 7]
```diff
@@ -408,35 +446,37 @@ export class Client {
...
+    if (useZstd) {
+      [marketBuffer, clockBuffer] =
+        await getConfirmedMarketsAndClockAccountsZstd(this.connection, [
+          marketKey,
+        ]);
+    } else {
+      [marketBuffer, clockBuffer] = await getConfirmedMarketsAndClockAccounts(
+        this.connection,
+        [marketKey]
+      );
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:473]: same

File: typescript/phoenix-sdk/src/client.ts [hunk 8]
```diff
@@ -344,48 +373,57 @@ export class Client {
...
+      accounts = await getConfirmedMarketsAndClockAccountsZstd(
+        this.connection,
+        marketKeys
+      );
+    } else {
+      accounts = await getConfirmedMarketsAndClockAccounts(
+        this.connection,
+        marketKeys
+      );
+    }
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/client.ts:424]: same

**Review by eugene-chen — APPROVED** [review]

File: typescript/phoenix-sdk/src/utils/connection.ts (hunk 6 above)
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/utils/connection.ts:109]: yeah i have no idea why :(

File: typescript/phoenix-sdk/src/client.ts (hunk 3 above)
Comment (eugene-chen) [reply typescript/phoenix-sdk/src/client.ts:343]: same everywhere we add this param

File: typescript/phoenix-sdk/src/utils/connection.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  connection: Connection,
+  marketKeys: PublicKey[]
+): Promise<Buffer[]> => {
+  const rawAccounts = (
+    await connection.getMultipleAccountsInfo(
+      [...marketKeys, SYSVAR_CLOCK_PUBKEY],
+      "confirmed"
+    )
+  ).map((account) => account?.data);
+  if (!rawAccounts.every((account) => account !== undefined)) {
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/connection.ts:22]: easier to read as:

`if (rawAccounts.some((account) => account === undefined))`

File: typescript/phoenix-sdk/src/utils/connection.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+): Promise<Buffer[]> => {
+  const body = {
+    jsonrpc: "2.0",
+    id: 1,
+    method: "getMultipleAccounts",
+    params: [
+      [...marketKeys.map((m) => m.toBase58()), SYSVAR_CLOCK_PUBKEY.toBase58()],
+      {
+        encoding: "base64+zstd",
+        commitment: "confirmed",
```
Comment (eugene-chen) [inline typescript/phoenix-sdk/src/utils/connection.ts:45]: This method and the one following should probably just take an optional CommitmentLevel that defaults to "confirmed"

File: typescript/phoenix-sdk/src/utils/connection.ts (hunk 9 above)
Comment (jarry-xiao) [reply typescript/phoenix-sdk/src/utils/connection.ts:22]: ah i was looking for the `any` function ty

**Pull request merged**